### PR TITLE
Add file system support to Jack compiler

### PR DESCRIPTION
**Pull request by netalondon**

### CODE REVIEW

File: components/src/stores/base/fs.ts [hunk 1]
```diff
@@ -26,7 +26,7 @@ export class FileSystemAccessFileSystemAdapter implements FileSystemAdapter {
   ): Promise<FileSystemDirectoryHandle> {
     let folder = this.baseDir;
     const parts = path.split(SEP).slice(1);
-    for (const part of parts) {
+    for (const part of parts.filter((part) => part.trim() != "")) {
```
Comment (DavidSouther) [inline components/src/stores/base/fs.ts:29]: There shouldn't be any empty directory pieces. If there are, the better way to handle it is a call to `Path::normalize`, but we don't have an easy to access copy of that for FS.

File: components/src/stores/compiler.store.ts [hunk 2]
```diff
@@ -22,19 +23,41 @@ export type CompilerStoreDispatch = Dispatch<{
   payload?: unknown;
 }>;
 
+function classTemplate(name: string) {
+  return `class ${name} {
```
Comment (DavidSouther) [inline components/src/stores/compiler.store.ts:27]: I think this would look better as

    return `class ${name} {\n\n}\n`;

File: web/public/pico.min.css
Comment (DavidSouther) [inline web/public/pico.min.css:1]: Revert this - there's a .prettierignore entry for it.

File: web/src/pages/compiler.tsx [hunk 3]
```diff
@@ -220,3 +207,60 @@ export default Compiler;
...
+}) => {
+  const [value, setValue] = useState<string>();
+
+  return (
+    <dialog open={dialog.isOpen}>
+      <article>
+        <header>
+          <Trans>{title}</Trans>
+          <a
+            style={{ color: "rgba(0, 0, 0, 0)" }}
```
Comment (DavidSouther) [inline web/src/pages/compiler.tsx:232]: Black transparent?

File: web/src/testing/index.tsx [hunk 4]
```diff
@@ -8,7 +8,7 @@ import { i18n } from "@lingui/core";
 import { I18nProvider } from "@lingui/react";
 import { render, RenderOptions, RenderResult } from "@testing-library/react";
 import ue from "@testing-library/user-event";
-import { ReactElement } from "react";
+import { ReactElement, createContext } from "react";
```
Comment (DavidSouther) [inline web/src/testing/index.tsx:11]: Remove

File: components/src/stores/compiler.store.ts [hunk 5]
```diff
@@ -22,19 +23,41 @@ export type CompilerStoreDispatch = Dispatch<{
...
+
+    setFile(
+      state: CompilerPageState,
+      { name, content }: { name: string; content: string }
+    ) {
+      state.files[name] = content;
+      this.compile(state);
+    },
+
     setFiles(state: CompilerPageState, files: Record<string, string>) {
```
Comment (DavidSouther) [inline components/src/stores/compiler.store.ts:54]: Let's add a comment that `files` needs to be the full path, not the basename, of the file.

**Review by DavidSouther — CHANGES_REQUESTED** [review]

File: components/src/stores/base/fs.ts (hunk 1 above)
Comment (DavidSouther) [reply components/src/stores/base/fs.ts:29]: In the mean time... move the `filter` to line 28 and we'll deal with it "later" :)

File: web/src/pages/compiler.tsx (hunk 3 above)
Comment (netalondon) [reply web/src/pages/compiler.tsx]: Oops, I guess that's what happens when you copy-paste. This was in settings as well.

**Pull request merged**